core/src/main/java/hudson/model/Slave.java

```diff
@@ -340,6 +344,23 @@ public String getLabelString() {
         return Util.fixNull(label).trim();
     }
 
+    private transient Set<LabelAtom> previouslyAssignedLabels = Collections.synchronizedSet(new HashSet<>());
```

Comment:
Or just
```suggestion
    private transient Set<LabelAtom> previouslyAssignedLabels = new HashSet<>();
```
(same in `readResolve`) and just make sure that `_setLabelString` synchronizes on it. Currently `drainLabelsToTrim` is acquiring a monitor so `synchronizedSet` is redundant for that access, which is harmless but a bit confusing.